### PR TITLE
Part 2 of the Host.Client split.

### PR DESCRIPTION
**Pull request by karthiknadig**
Adds a bunch of services to support WebSocketClient, Remote help and shiny, and assembly by type.

### CODE REVIEW

**Comment by karthiknadig** [comment]
I need to add the copyright header.  I will send a update with that.

File: src/Host/Client/Impl/BrokerServices/WebService.cs [hunk 1]
```diff
@@ -81,17 +81,18 @@ private Task RepeatUntilAuthenticatedAsync(Func<CancellationToken, Task> action,
             }
         }
 
-        public Task<TResponse> HttpGetAsync<TResponse>(UriTemplate uriTemplate, CancellationToken cancellationToken = default(CancellationToken), params object[] args) =>
-           HttpGetAsync<TResponse>(MakeUri(uriTemplate, args), cancellationToken);
+        //public Task<TResponse> HttpGetAsync<TResponse>(UriTemplate uriTemplate, CancellationToken cancellationToken = default(CancellationToken), params object[] args) =>
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/BrokerServices/WebService.cs]: Is it not needed anymore?

File: src/Host/Client/Impl/Extensions/AboutHostExtensions.cs [hunk 2]
```diff
@@ -2,18 +2,19 @@
...
 
 using System;
+using System.Reflection;
 using Microsoft.Common.Core;
 using Microsoft.R.Host.Protocol;
 
 namespace Microsoft.R.Host.Client {
     public static class AboutHostExtensions {
-        private static readonly Version _localVersion;
+        private static Version _localVersion;
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Extensions/AboutHostExtensions.cs]:
```cs
static AboutHostExtensions() {
    _localVersion = typeof(AboutHost).GetTypeInfo().Assembly.GetName().Version;
}
```

See `AboutController` - it does exactly the same

File: src/Host/Client/Impl/Extensions/AboutHostExtensions.cs [hunk 3]
```diff
@@ -2,18 +2,19 @@
...
 
-        static AboutHostExtensions() {
-            _localVersion = typeof(AboutHost).Assembly.GetName().Version;
-        }
 
-        public static string IsHostVersionCompatible(this AboutHost aboutHost) {
+        public static string IsHostVersionCompatible(this AboutHost aboutHost, Assembly aboutHostAssembly) {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Extensions/AboutHostExtensions.cs]: `Assembly aboutHostAssembly` isn't needed

File: src/Host/Client/Impl/Host/RemoteBrokerClient.cs [hunk 4]
```diff
@@ -26,12 +28,14 @@ internal sealed class RemoteBrokerClient : BrokerClient {
         private bool? _certificateValidationResult;
 
         static RemoteBrokerClient() {
+#if !NETSTANDARD1_6
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Host/RemoteBrokerClient.cs]: These two should come in 2.0 
Can we survive without them?

File: src/Host/Client/Impl/Host/RemoteBrokerClient.cs [hunk 5]
```diff
@@ -26,12 +28,14 @@ internal sealed class RemoteBrokerClient : BrokerClient {
...
             ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
             ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
+#endif
         }
 
         public RemoteBrokerClient(string name, BrokerConnectionInfo connectionInfo, IServiceContainer services, IConsole console, CancellationToken cancellationToken)
-            : base(name, connectionInfo, new RemoteCredentialsDecorator(connectionInfo.CredentialAuthority, connectionInfo.Name, services), services.Log(), console) {
+            : base(name, connectionInfo, new RemoteCredentialsDecorator(connectionInfo.CredentialAuthority, connectionInfo.Name, services), services.Log(), console, services) {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Host/RemoteBrokerClient.cs]: No reason to pass `services.Log()` if you pass `services`

File: src/Host/Client/Impl/Session/RSession.cs [hunk 6]
```diff
@@ -83,7 +85,7 @@ internal sealed class RSession : IRSession, IRCallbacks {
             CanceledBeginInteractionTask = TaskUtilities.CreateCanceled<IRSessionInteraction>(new RHostDisconnectedException());
         }
 
-        public RSession(int id, string name, IBrokerClient brokerClient, IExclusiveReaderLock initializationLock, Action onDispose) {
+        public RSession(int id, string name, IBrokerClient brokerClient, IExclusiveReaderLock initializationLock, Action onDispose, IServiceContainer services) {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSession.cs]: `RSession` shouldn't have access to the `IServiceContainer` directly, it should interact with external world only through the callback. Why do you need it?

File: src/Host/Client/Impl/Session/RSession.cs [hunk 7]
```diff
@@ -392,9 +395,10 @@ public Task StartHostAsync(RHostStartupInfo startupInfo, IRSessionCallback callb
         private async Task AfterHostStarted(RHostStartupInfo startupInfo) {
             var evaluator = new BeforeInitializedRExpressionEvaluator(this);
             try {
+                var rHostAssembly = _services.GetService<ILocalClientServices>().GetAssemblyByType(typeof(RHost));
                 // Load RTVS R package before doing anything in R since the calls
                 // below calls may depend on functions exposed from the RTVS package
-                var libPath = IsRemote ? "." : Path.GetDirectoryName(Assembly.GetExecutingAssembly().GetAssemblyPath());
+                var libPath = IsRemote ? "." : Path.GetDirectoryName(rHostAssembly.GetAssemblyPath());
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSession.cs]: `typeof(RSession)..GetTypeInfo().Assembly.GetAssemblyPath()`

File: src/Host/Client/Impl/Transports/IWebSocketClient.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Net.WebSockets;
+using System.Threading.Tasks;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.R.Host.Client {
+    public interface IWebSocketClient {
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Transports/IWebSocketClient.cs]: Are you planning to move `WebSocketClient` to xplat as well?

File: src/Windows/Host/Client/Impl/Host/LocalClientServices.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.R.Host.Client {
+    public class LocalClientServices : ILocalClientServices {
```
Comment (AlexanderSher) [inline src/Windows/Host/Client/Impl/Host/LocalClientServices.cs]: You don't need this service

File: src/Host/Client/Impl/BrokerServices/WebService.cs (hunk 1 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/BrokerServices/WebService.cs]: nope

---
In reply to: [111815999](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111815999) [](ancestors = 111815999)

File: src/Host/Client/Impl/Extensions/AboutHostExtensions.cs (hunk 2 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/Extensions/AboutHostExtensions.cs]: Updated and remove the service i had added to handle this.

---
In reply to: [111816503](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111816503) [](ancestors = 111816503)

File: src/Host/Client/Impl/Session/RSession.cs (hunk 6 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/Session/RSession.cs]: removed the service requirement.

---
In reply to: [111820325](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111820325) [](ancestors = 111820325)

File: src/Windows/Host/Client/Impl/Host/LocalClientServices.cs (hunk 9 above)
Comment (karthiknadig) [inline src/Windows/Host/Client/Impl/Host/LocalClientServices.cs]: agreed and removed.

---
In reply to: [111822088](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111822088) [](ancestors = 111822088)

File: src/Host/Client/Impl/Transports/IWebSocketClient.cs (hunk 8 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/Transports/IWebSocketClient.cs]: yes. eventually. There are a few authentication related questions regarding wss (web socket secure).

---
In reply to: [111821736](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111821736) [](ancestors = 111821736)

File: src/Host/Client/Impl/Session/RSession.cs (hunk 7 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/Session/RSession.cs]: fixed

---
In reply to: [111820628](https://github.com/Microsoft/RTVS/pull/3419#discussion_r111820628) [](ancestors = 111820628)

File: src/Host/Client/Impl/Host/RemoteBrokerClient.cs (hunk 4 above)
Comment (karthiknadig) [inline src/Host/Client/Impl/Host/RemoteBrokerClient.cs]: @MikhailArkhipov any suggestions about this?

**Comment by AlexanderSher** [comment]
👍

**Comment by MikhailArkhipov** [comment]
👍

**Pull request merged**